desc/magnetic_fields.py

```diff
@@ -795,3 +831,265 @@ def odefun(rpz, s):
+                    "May cause errors unless potential function is also changed."
+                )
+            self._params = new
+            # recompute K if params changed
+            self._compute_surface_current()
+
+    @property
+    def potential(self):
+        """Potential function, signature (theta,zeta,**params) -> potential value."""
+        return self._potential
+
+    @potential.setter
+    def potential(self, new):
+        if new != self._potential:
+            self._potential = new
+            # reset derivatives if potential has changed, using AD
+            self._set_derivatives()
+            # recompute K if potential changed
+            self._compute_surface_current()
+
+    def _set_derivatives(self, potential_dtheta=None, potential_dzeta=None):
+        if potential_dtheta:
+            self._potential_t = partial(potential_dtheta, **self.params)
+        else:
+            self._potential_t = vmap(
+                Derivative(
+                    partial(self.potential, **self.params),
+                    argnum=0,
+                    mode="grad",
+                ).compute,
+                in_axes=[0, 0],
+            )
+        if potential_dzeta:
+            self._potential_z = partial(potential_dzeta, **self.params)
+        else:
+            self._potential_z = vmap(
+                Derivative(
+                    partial(self.potential, **self.params), argnum=1, mode="grad"
+                ).compute,
+                in_axes=[0, 0],
+            )
+
+    def _compute_surface_current(self, surface_grid=None, params=None):
+        if surface_grid is None:
+            surface_grid = self.surface_grid
+        if (params is None) or (len(params) == 0):
+            params = self._params
+
+        # surface is the source of current density for the magnetic field
+        # compute source positions rs, and their theta/zeta derivatives
+        # store in "rpz" basis so can rotate them during compute_magnetic_field
+        data = self.compute(["x", "e_theta", "e_zeta"], grid=surface_grid, basis="xyz")
+        self._rs = xyz2rpz_vec(data["x"], phi=self.surface_grid.nodes[:, 2])
+        theta = surface_grid.nodes[:, 1]
+        zeta = surface_grid.nodes[:, 2]
+        # compute the dV element for the surface needed for biot savart
+        self._dV = (
+            surface_grid.weights
+            * jnp.linalg.norm(
+                jnp.cross(data["e_theta"], data["e_zeta"], axis=-1), axis=-1
+            )
+            / surface_grid.NFP
+        )  # divide by NFP here so that the NFP loop
+        # in compute_magnetic_field is correct
+
+        # compute the potential derivatives on the surface
+        phi_t = self._potential_t(theta, zeta)
+        phi_z = self._potential_z(theta, zeta)
+
+        # compute surface normal magnitude
+        ns_mag = jnp.linalg.norm(jnp.cross(data["e_theta"], data["e_zeta"]), axis=1)
```

Comment:
if you end up pushing another commit you could also move ns above self._dv and use it for the computation there

Reply:
smart, I can do that